relay-event-normalization/src/normalize/span/description/mod.rs

```diff
@@ -279,25 +278,51 @@ pub fn scrub_ipv6(ip: Ipv6Addr) -> &'static str {
 /// use relay_event_normalization::span::description::scrub_domain_name;
 ///
 /// assert_eq!(scrub_domain_name("my.domain.com"), "*.domain.com");
+/// assert_eq!(scrub_domain_name("data.bbc.co.uk"), "*.bbc.co.uk");
 /// assert_eq!(scrub_domain_name("hello world"), "hello world");
 /// ```
 pub fn scrub_domain_name(domain: &str) -> Cow<'_, str> {
     if DOMAIN_ALLOW_LIST.contains(&domain) {
         return Cow::Borrowed(domain);
     }
 
-    let mut tokens = domain.rsplitn(3, '.');
-    let tld = tokens.next();
-    let domain = tokens.next();
-    let prefix = tokens.next().map(|_| "*");
-
-    Cow::Owned(
-        prefix
-            .iter()
-            .chain(domain.iter())
-            .chain(tld.iter())
-            .join("."),
-    )
+    let parsed_domain = psl::domain(domain.as_bytes());
+
+    if parsed_domain.is_none() {
+        // If parsing fails, return the original string
+        return Cow::Borrowed(domain);
+    }
+
+    let parsed_domain = parsed_domain.unwrap();
+
+    let suffix_string = String::from_utf8(parsed_domain.suffix().as_bytes().to_vec());
+    let scrubbed_suffix = match &suffix_string {
+        Ok(string) => string.as_ref(),
+        Err(_) => "*",
+    };
+
+    let domain_string = String::from_utf8(parsed_domain.as_bytes().to_vec());
+    let scrubbed_domain = match &domain_string {
+        Ok(string) => {
+            let second_level_domain = string.strip_suffix(scrubbed_suffix).unwrap_or("");
+
+            // If a subdomain is present, replace with `"*."`
+            let subdomain = match domain
+                .strip_suffix(scrubbed_suffix)
+                .unwrap_or(domain)
+                .strip_suffix(second_level_domain)
+            {
+                None => "",
+                Some("") => "",
+                _ => "*.",
+            };
+
+            format!("{subdomain}{second_level_domain}")
+        }
+        Err(_) => String::from("*"),
+    };
+
+    return Cow::Owned(format!("{scrubbed_domain}{scrubbed_suffix}"));
```

Comment:
I gave this a shot, not sure if it's more readable but it uses fewer string allocations:

```suggestion
    let suffix = parsed_domain.suffix().as_bytes();
    let Some(second_level_domain) = parsed_domain.as_bytes().strip_suffix(suffix) else {
        return Cow::Borrowed(domain);
    };

    let subdomain = domain
        .as_bytes()
        .strip_suffix(suffix)
        .and_then(|s| s.strip_suffix(second_level_domain));

    match subdomain {
        None | Some(b"") => Cow::Borrowed(domain),
        Some(_subdomain) => {
            let scrubbed = [b"*.", second_level_domain, suffix].concat();
            match String::from_utf8(scrubbed) {
                Ok(s) => Cow::Owned(s),
                Err(_) => Cow::Borrowed("*"),
            }
        }
    }
```